Mathlib/AlgebraicGeometry/EllipticCurve/Weierstrass.lean

```diff
@@ -532,117 +526,127 @@ section ModelsWithJ
-  field_simp [EllipticCurve.j, ofJ', @val_unitOfInvertible _ _ _ <| invertibleMul ..,
-    WeierstrassCurve.ofJ_c₄]
+It is of j-invariant j (see `WeierstrassCurve.ofJNe0Or1728_j`). -/
+instance (j : R) [h1 : Fact (IsUnit j)] [h2 : Fact (IsUnit (j - 1728))] :
+    (ofJNe0Or1728 j).IsElliptic := by
+  rw [isElliptic_iff, ofJNe0Or1728_Δ j]
+  exact (h1.out.pow 2).mul (h2.out.pow 9)
+
+-- TODO: change to `[IsUnit ...]` once #17458 is merged
+lemma ofJNe0Or1728_j (j : R) [Fact (IsUnit j)] [Fact (IsUnit (j - 1728))] :
+    (ofJNe0Or1728 j).j = j := by
+  field_simp [WeierstrassCurve.j, ofJNe0Or1728_c₄, ofJNe0Or1728_Δ]
   ring1
 
 variable {F : Type u} [Field F] (j : F)
 
-private lemma two_or_three_ne_zero : (2 : F) ≠ 0 ∨ (3 : F) ≠ 0 :=
-  ne_zero_or_ne_zero_of_nat_coprime <| by decide
-
 variable [DecidableEq F]
 
 /-- For any element j of a field $F$, there exists an elliptic curve over $F$
-with j-invariant equal to j (see `EllipticCurve.ofJ_j`).
-Its coefficients are given explicitly (see `EllipticCurve.ofJ0`, `EllipticCurve.ofJ1728`
-and `EllipticCurve.ofJ'`). -/
-def ofJ : EllipticCurve F :=
-  if h0 : j = 0 then
-    if h3 : (3 : F) = 0 then @ofJ1728 _ _ <| invertibleOfNonzero <|
-      two_or_three_ne_zero.neg_resolve_right h3
-    else @ofJ0 _ _ <| invertibleOfNonzero h3
-  else if h1728 : j = 1728 then
-    @ofJ1728 _ _ <| invertibleOfNonzero fun h => h0 <|
-    by rw [h1728, show (1728 : F) = 2 * 864 by norm_num1, h, zero_mul]
-  else @ofJ' _ _ j (invertibleOfNonzero h0) (invertibleOfNonzero <| sub_ne_zero_of_ne h1728)
-
-lemma ofJ_0_of_three_ne_zero [h3 : NeZero (3 : F)] :
-    ofJ 0 = @ofJ0 _ _ (invertibleOfNonzero h3.out) := by
-  rw [ofJ, dif_pos rfl, dif_neg h3.out]
-
-lemma ofJ_0_of_three_eq_zero (h3 : (3 : F) = 0) :
-    ofJ 0 = @ofJ1728 _ _ (invertibleOfNonzero <| two_or_three_ne_zero.neg_resolve_right h3) := by
-  rw [ofJ, dif_pos rfl, dif_pos h3]
-
-lemma ofJ_0_of_two_eq_zero (h2 : (2 : F) = 0) :
-    ofJ 0 = @ofJ0 _ _ (invertibleOfNonzero <| two_or_three_ne_zero.neg_resolve_left h2) :=
-  have := neZero_iff.2 <| two_or_three_ne_zero.neg_resolve_left h2
-  ofJ_0_of_three_ne_zero
-
-lemma ofJ_1728_of_three_eq_zero (h3 : (3 : F) = 0) :
-    ofJ 1728 = @ofJ1728 _ _ (invertibleOfNonzero <| two_or_three_ne_zero.neg_resolve_right h3) := by
-  rw [ofJ, dif_pos <| by rw [show (1728 : F) = 3 * 576 by norm_num1, h3, zero_mul], dif_pos h3]
-
-lemma ofJ_1728_of_two_ne_zero [h2 : NeZero (2 : F)] :
-    ofJ 1728 = @ofJ1728 _ _ (invertibleOfNonzero h2.out) := by
+with j-invariant equal to j (see `WeierstrassCurve.ofJ_j`).
+Its coefficients are given explicitly (see `WeierstrassCurve.ofJ0`, `WeierstrassCurve.ofJ1728`
+and `WeierstrassCurve.ofJNe0Or1728`). -/
+def ofJ : WeierstrassCurve F :=
+  if j = 0 then if (3 : F) = 0 then ofJ1728 F else ofJ0 F
+  else if j = 1728 then ofJ1728 F else ofJNe0Or1728 j
+
+lemma ofJ_0_of_three_ne_zero [h3 : NeZero (3 : F)] : ofJ 0 = ofJ0 F := by
```

Comment:
Should we use `(h3 : (3 : F) ≠ 0)` instead? I don't quite remember why we did this instead.

Otherwise, we can also make the assumptions below `[CharP F 3]` etc, but we should be consistent.

Reply:
> Should we use `(h3 : (3 : F) ≠ 0)` instead? I don't quite remember why we did this instead.

It's here https://github.com/leanprover-community/mathlib4/pull/5935#discussion_r1280637704

I think it's OK to revert to `(h3 : (3 : F) ≠ 0)` since otherwise it looks like that we are abusing instances.